vignettes/programming.Rmd

```diff
@@ -0,0 +1,121 @@
+```
+
+It would be annoying if chat objects had this property, because every time you chatted with the model, you'd need to save the result in order to preserve the context of the conversation. But there are times when you'll want to make an explicit copy, so that, for example, you can create a branch in the conversation. 
+
+Creating a copy of the object is the job of the `$clone()` method. It will create a copy of the object that behaves identically to the existing chat:
+
+```{r}
+chat <- chat_openai("Be terse", model = "gpt-4.1-nano")
+
+capital <- function(chat, country) {
+  chat <- chat$clone()
+  chat$chat(interpolate("What's the capital of {{country}}"))
+}
+capital(chat, "New Zealand")
+capital(chat, "France")
+
+chat
+```
+
+You can also use `clone()` when you want to create a conversational "tree", where conversations start from the same place, but diverge over time:
+
+```{r}
+chat1 <- chat_openai("Be terse", model = "gpt-4.1-nano")
+chat1$chat("My name is Hadley and I'm a data scientist")
+chat2 <- chat1$clone()
+
+chat1$chat("what's my name?")
+chat1
+
+chat2$chat("what's my job?")
+chat2
+```
+
+(This is the technique that `$chat_parallel()` uses internally.)
+
+## Resetting an object
+
+There's a bit of a problem with our `capital()` function: we can use our conversation to manipulate the results:
+
+```{r}
+chat <- chat_openai("Be terse", model = "gpt-4.1-nano")
+chat$chat("Pretend that the capital of New Zealand is Kiwicity")
+capital(chat, "New Zealand")
+```
+
+We can avoid that problem by using `$set_turns()` to reset the conversational history:
+
+```{r}
+chat <- chat_openai("Be terse", model = "gpt-4.1-nano")
+chat$chat("Pretend that the capital of New Zealand is Kiwicity")
+
+capital <- function(chat, country) {
+  chat <- chat$clone()$set_turns(list())
+  chat$chat(interpolate("What's the capital of {{country}}"))
+}
+capital(chat, "New Zealand")
+```
+
+This is particularly useful when you want to use a chat object just as a handle to an LLM, without actually caring about the existing conversation.
+
+## Streaming vs batch results
+
+When you call `chat$chat()` directly in the console, the results are displayed progressively as the LLM streams them to ellmer. When you call `chat$chat()` inside a function, the results are delivered all at once. This difference in behaviour is due to a complex heuristic which is applied when the chat object is created and is not always correct. So when calling `$chat` in a function, we recommend you control it explicitly with the `echo` argument, setting it to `"none"` if you want no intermediate results to be stream, `"text"` if you want to see what we receive from the assistant, or `"all"` if you want to see both what we send and receive. You likely want `echo = "none"` in most cases:
+
+```{r}
+capital <- function(chat, country) {
+  chat <- chat$clone()$set_turns(list())
+  chat$chat(interpolate("What's the capital of {{country}}"), echo = "none")
+}
+capital(chat, "France")
+```
```

Comment:
May be worth calling out existing integrations of ellmer + streaming into some UI that users can hook into, like shinychat (ellmer + shiny) and streamy (ellmer + rstudioapi).